File: day2/solution2.1.py
def is_safe_report(report):
    """
    Determine if a report is safe.
    """
    differences = [report[i + 1] - report[i] for i in range(len(report) - 1)]

    # Check if all differences are valid (between 1 and 3 or -1 and -3)
    if not all(1 <= abs(diff) <= 3 for diff in differences):
        return False

    # Check if the report is strictly increasing or strictly decreasing
    if all(diff > 0 for diff in differences) or all(diff < 0 for diff in differences):
        return True

    return False


def count_safe_reports(file_path):
    """
    Count the number of safe reports in the input file.
    """
    with open(file_path, 'r') as file:
        lines = file.readlines()

    safe_count = 0
    for line in lines:
        report = list(map(int, line.strip().split()))
        if is_safe_report(report):
            safe_count += 1

    return safe_count


if __name__ == "__main__":
    input_file = "input.txt"
    safe_reports = count_safe_reports(input_file)
    print(f"Number of safe reports: {safe_reports}")

File: day2/solution2.1.ts
import * as fs from "fs";

function isSafeReport(report: number[]): boolean {
    // Calculate the differences between adjacent levels
    const differences = report.slice(1).map((level, i) => level - report[i]);

    // Check if all differences are valid (between 1 and 3 or -1 and -3)
    if (!differences.every(diff => Math.abs(diff) >= 1 && Math.abs(diff) <= 3)) {
        return false;
    }

    // Check if the report is strictly increasing or strictly decreasing
    const allIncreasing = differences.every(diff => diff > 0);
    const allDecreasing = differences.every(diff => diff < 0);

    return allIncreasing || allDecreasing;
}

function countSafeReports(filePath: string): number {
    // Read and parse the input file
    const data = fs.readFileSync(filePath, "utf-8");
    const lines = data.trim().split("\n");

    let safeCount = 0;
    for (const line of lines) {
        const report = line.trim().split(" ").map(Number);
        if (isSafeReport(report)) {
            safeCount++;
        }
    }

    return safeCount;
}

// Main Execution
const inputFile = "input.txt";
const safeReports = countSafeReports(inputFile);
console.log(`Number of safe reports: ${safeReports}`);